pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,185 @@
+package remotecluster
+
+import (
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	v1net "k8s.io/api/networking/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+type RemoteCluster struct {
+	stopCh chan struct{}
+
+	// notificationMutex is used to block incoming events during an
+	// OnSyncDone event being sent back to the controller
+	ClusterID             string
+	ClientSet             kubernetes.Interface
+	podInformer           cache.SharedIndexInformer
+	networkPolicyInformer cache.SharedIndexInformer
+
+	eventChan chan Event
+}
+
+const AddEvent = "Added"
+const DeleteEvent = "Deleted"
+const UpdateEvent = "Updated"
+
+const NetworkPolicy = "np"
+const Pod = "pod"
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    string
+	ObjType string
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) SetEventChannel(eventChan chan Event) {
+	rw.eventChan = eventChan
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rw.stopCh, rw.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rw.stopCh, rw.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rw)
+		}
```

Comment:
To run it in parallel and not block on run, then get the callback asynchronously later. Another approach would be not to parallelize, remove the callback, and asume this will exit when on sync.

But I wanted to speedup the sync to all discovered clusters at the start. Does that make sense?

Or do you mean I should write a function and not use the anonymous of this code?

Reply:
Sorry, as above I misread that it's a go-routine. It makes sense to do it async.